source/common/config/datasource.cc

```diff
@@ -61,12 +43,109 @@ absl::StatusOr<std::string> read(const envoy::config::core::v3::DataSource& sour
   return data;
 }
 
+absl::StatusOr<std::string> readFile(const std::string& path, Api::Api& api, bool allow_empty,
+                                     uint64_t max_size) {
+  auto& file_system = api.fileSystem();
+
+  if (max_size > 0) {
+    if (!file_system.fileExists(path)) {
+      return absl::InvalidArgumentError(fmt::format("file {} does not exist", path));
+    }
+    const ssize_t size = file_system.fileSize(path);
+    if (size < 0) {
+      return absl::InvalidArgumentError(absl::StrCat("cannot determine size of file ", path));
+    }
+    if (static_cast<uint64_t>(size) > max_size) {
+      return absl::InvalidArgumentError(
+          fmt::format("file {} size is {} bytes; maximum is {}", path, size, max_size));
+    }
+  }
+
+  auto file_content_or_error = file_system.fileReadToEnd(path);
+  RETURN_IF_STATUS_NOT_OK(file_content_or_error);
+
+  if (!allow_empty && file_content_or_error.value().empty()) {
+    return absl::InvalidArgumentError(fmt::format("file {} is empty", path));
+  }
+
+  return file_content_or_error.value();
+}
+
 absl::optional<std::string> getPath(const envoy::config::core::v3::DataSource& source) {
   return source.specifier_case() == envoy::config::core::v3::DataSource::SpecifierCase::kFilename
              ? absl::make_optional(source.filename())
              : absl::nullopt;
 }
 
+DynamicData::DynamicData(Event::Dispatcher& main_dispatcher,
+                         ThreadLocal::TypedSlotPtr<ThreadLocalData> slot,
+                         Filesystem::WatcherPtr watcher)
+    : dispatcher_(main_dispatcher), slot_(std::move(slot)), watcher_(std::move(watcher)) {}
+
+DynamicData::~DynamicData() {
+  if (!dispatcher_.isThreadSafe()) {
+    dispatcher_.post([to_delete = std::move(slot_)] {});
```

Comment:
Do we need to delete the thread local on the main? I added something very similar https://github.com/envoyproxy/envoy/blob/main/source/common/secret/secret_provider_impl.h#L124, that doesn't delete it on main.

Reply:
Yeah, all slot should be deleted on the main. But most loaded configuration (and their thread local storage) will be  destroyed on main.
The data source provider here may be used in route which may be destroyed on workers, so I added the post() here.

See also https://github.com/envoyproxy/envoy/pull/33395